File: core/src/main/java/org/predicode/predicator/grammar/TermPrinter.java
package org.predicode.predicator.grammar;

import org.predicode.predicator.terms.Term;

import javax.annotation.Nonnull;

import static org.predicode.predicator.grammar.CodePoints.*;
import static org.predicode.predicator.grammar.QuotingStyle.AUTO_QUOTE;
import static org.predicode.predicator.grammar.QuotingStyle.OPEN_QUOTE;
import static org.predicode.predicator.grammar.TermSep.*;


public class TermPrinter {

    public static void printTerms(@Nonnull Iterable<? extends Term> terms, @Nonnull CodePointPrinter print) {
        new TermPrinter(print).print(terms);
    }

    @Nonnull
    public static String printTerms(@Nonnull Iterable<? extends Term> terms) {

        final StringBuilder out = new StringBuilder();

        printTerms(terms, out::appendCodePoint);

        return out.toString();
    }

    public static void printTerms(@Nonnull CodePointPrinter print, @Nonnull Term ...terms) {
        new TermPrinter(print).print(terms);
    }

    @Nonnull
    public static String printTerms(@Nonnull Term ...terms) {

        final StringBuilder out = new StringBuilder();

        printTerms(out::appendCodePoint, terms);

        return out.toString();
    }

    @Nonnull
    private final CodePointPrinter out;

    @Nonnull
    private TermSep sep = NO_SEP;

    private TermPrinter(@Nonnull CodePointPrinter out) {
        this.out = out;
    }

    public void print(@Nonnull Iterable<? extends Term> terms) {
        for (Term term : terms) {
            term.print(this);
        }
    }

    public void print(@Nonnull Term ...terms) {
        for (Term term : terms) {
            term.print(this);
        }
    }

    public void keyword(@Nonnull CharSequence name, @Nonnull QuotedName quoted) {

        final QuotingStyle quoting;

        if (quoted.isInfix()) {
            quoting = this.sep.infix(this.out);
        } else {
            this.sep.keyword(this.out);
            quoting = AUTO_QUOTE;
        }

        quoting.printName(name, BACKTICK, this.out);
        this.sep = quoted.isPrefix() ? PREFIX_SEP : KEYWORD_SEP;
    }

    public void atom(@Nonnull CharSequence name) {
        quotedName(name, ATOM_SEP);
    }

    public void variable(@Nonnull CharSequence name) {
        quotedName(name, VARIABLE_SEP);
    }

    private void quotedName(@Nonnull CharSequence name, @Nonnull TermSep.QuotedSep sep) {
        this.sep.quoted(this.out);

        final boolean quoteClosed = OPEN_QUOTE.printName(name, sep.getQuote(), this.out);

        this.sep = quoteClosed ? SPACE_SEP : sep;
    }

    public void value(@Nonnull CharSequence value) {
        this.sep.value(this.out);
        this.out.print(OPENING_BRACE);
        this.out.print(value);
        this.out.print(CLOSING_BRACE);
        this.sep = SPACE_SEP;
    }

    public void special(@Nonnull CharSequence value) {
        this.sep.special(this.out);
        this.out.print(value);
        this.sep = SPECIAL_SEP;
    }

    public void startCompound() {
        this.sep.value(this.out);
        this.out.print(OPENING_PARENT);
        this.sep = NO_SEP;
    }

    public void endCompound() {
        this.out.print(CLOSING_PARENT);
        this.sep = SPACE_SEP;
    }

}

File: core/src/main/kotlin/org/predicode/predicator/grammar/TermPrinter.kt
package org.predicode.predicator.grammar

import org.predicode.predicator.Term

class TermPrinter internal constructor(private val print: CodePointConsumer) {

    private var partPrinter: PartPrinter = InitialTermPrinter

    fun print(terms: Iterable<Term>) {
        terms.forEach { term -> term.print(this) }
    }

    fun print(vararg terms: Term) {
        terms.forEach { term -> term.print(this) }
    }

    fun keyword(name: CharSequence) {
        partPrinter = partPrinter.keyword(this, name)
    }

    fun atom(name: CharSequence) {
        this.partPrinter = quotedName(name, quote = SINGLE_QUOTE)
    }

    fun variable(name: CharSequence) {
        this.partPrinter = quotedName(name, quote = UNDERSCORE)
    }

    private fun quotedName(name: CharSequence, quote: CodePoint): PartPrinter =
            this.partPrinter.separate(this).let {

                val quoteClosed = printName(name, quote = quote, quoting = QuotingStyle.OPEN_QUOTE, print = print)

                if (quoteClosed) return UnquotedTermPrinter

                return it.quoted(quote)
            }

    fun value(value: CharSequence) {
        this.partPrinter = this.partPrinter.separate(this).apply {
            out(OPENING_BRACE)
            value.codePoints().forEach { out(it) }
            out(CLOSING_BRACE)
        }
    }

    fun startCompound() {
        this.partPrinter = this.partPrinter.endQuoted(this).separate(this).let {
            out(OPENING_PARENT)
            InitialTermPrinter
        }
    }

    fun endCompound() {
        this.partPrinter = this.partPrinter.endQuoted(this).apply {
            out(CLOSING_PARENT)
        }
    }

    private fun out(codePoint: CodePoint) = print(codePoint)

    private interface PartPrinter {

        fun separate(printer: TermPrinter): PartPrinter

        fun endQuoted(printer: TermPrinter): PartPrinter

        @JvmDefault
        fun keyword(printer: TermPrinter, name: CharSequence): PartPrinter =
                endQuoted(printer).separate(printer).apply {
                    printName(name, quote = BACKTICK, print = printer.print)
                }

        @JvmDefault
        fun quoted(quote: CodePoint): PartPrinter =
                QuotedTermPrinter(quote)

    }

    private object InitialTermPrinter : PartPrinter {

        override fun separate(printer: TermPrinter) = UnquotedTermPrinter

        override fun endQuoted(printer: TermPrinter) = this

    }

    private class QuotedTermPrinter(val quote: CodePoint) : PartPrinter {

        override fun separate(printer: TermPrinter) = endQuoted(printer).apply { printer.out(SPACE) }

        override fun keyword(printer: TermPrinter, name: CharSequence): PartPrinter {
            printer.out(quote)
            return super.keyword(printer, name).endQuoted(printer)
        }

        override fun endQuoted(printer: TermPrinter) = UnquotedTermPrinter

    }

    private object UnquotedTermPrinter : PartPrinter {

        override fun separate(printer: TermPrinter) = apply { printer.out(SPACE) }

        override fun endQuoted(printer: TermPrinter) = this

    }

}

fun printTerms(terms: Iterable<Term>, print: CodePointConsumer) {
    TermPrinter(print).print(terms)
}

fun printTerms(terms: Iterable<Term>): String = StringBuilder().apply {
    printTerms(terms) { appendCodePoint(it) }
}.toString()

fun printTerms(vararg terms: Term, print: CodePointConsumer) {
    TermPrinter(print).print(*terms)
}

fun printTerms(vararg terms: Term): String = StringBuilder().apply {
    printTerms(*terms) { appendCodePoint(it) }
}.toString()
